### PR TITLE
Jelmer hud final

### PR DESCRIPTION
**Pull request by waseland**

### CODE REVIEW

**Comment by waseland** [comment]
Fuck this shit, fucking rebasing shitstorm of commits, I only used git pull --rebase and rebased using the mergetool and I still get this shit :( Well after 2 hours everything is working, I can confirm that but the PR is total bullshit now :/

File: src/gamelogic/HUD.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    using Engine::Updatable;
+    using Engine::Color;
+    using Engine::SDLFacade;
+    using Engine::FontType ;
+
+    class HUD : public Drawable, public Updatable {
+    public:
+        HUD(SDLFacade& _SDL_facade, Player& player);
+
+        int get_current_time();
```
Comment (jorgdebont) [inline src/gamelogic/HUD.hpp]: Function is not used

File: src/gamelogic/Player.hpp [hunk 2]
```diff
@@ -47,9 +49,11 @@ namespace GameLogic {
         double _rotation = 0;
         double _next_tile = 0.5;
         bool _action_released = true;
-
+
```
Comment (jorgdebont) [inline src/gamelogic/Player.hpp]: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAaaaa

File: src/gamelogic/state/PauseState.cpp [hunk 3]
```diff
@@ -16,7 +16,7 @@ namespace State {
         MenuOption resume_game {
             {150,100},
             "Resume game",
-            [] (GameLogic::Game& game) {
+            [=] (GameLogic::Game& game) {
```
Comment (jorgdebont) [inline src/gamelogic/state/PauseState.cpp]: Only capture what you need

File: src/gamelogic/state/RunState.cpp [hunk 4]
```diff
@@ -27,8 +32,10 @@ namespace State {
         Input keys = this->_context.SDL_facade.get_input();
         for (auto key : keys.keys_released) {
             switch (key) {
-                case Key::ESC:
+                case Key::ESC: {
```
Comment (jorgdebont) [inline src/gamelogic/state/RunState.cpp]: why { and }

File: src/gamelogic/HUD.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            _SDL_facade.draw_image("res/look_front.bmp", this->_character_head_pos);
+        } else {
+            _SDL_facade.draw_image("res/look_right.bmp", this->_character_head_pos);
+        }
+    }
+
+    void HUD::draw()
+    {
+        //draw HUD
+        _SDL_facade.draw_image("res/HUDv2.bmp", this->_HUD_pos);
```
Comment (jorgdebont) [inline src/gamelogic/HUD.cpp]: hardcoded path

File: src/gamelogic/HUD.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        this->_current_ticks = _SDL_facade.get_ticks() - this->_start_ticks;
+        this->_current_time = this->_calculate_time(this->_current_ticks);
+    }
+
+    int HUD::_calculate_time(int ticks)
+    {
+        return (ticks / 1000);
+    }
+
+    string HUD::_time_to_string(int current_time)
```
Comment (jorgdebont) [inline src/gamelogic/HUD.cpp]: you could also use string streams for this, but i dont really care

File: src/gamelogic/state/RunState.cpp (hunk 4 above)
Comment (SandervGeel) [inline src/gamelogic/state/RunState.cpp]: was nodig toen er nog ticks mee gegeven werden, zijn nu weg gehaald

File: src/gamelogic/HUD.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    void HUD::set_current_tick(int ticks)
+    {
+        this->_current_ticks = ticks;;
+    }
+
+    int HUD::_calculate_health_blocks(int health)
+    {
+        return (int) ceil((double)health / (this->_total_health / this->_amount_health_blocks));
```
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: done

File: src/gamelogic/HUD.cpp [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    void HUD::_draw_health_text(int health)
```
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: done :dog:

File: src/gamelogic/HUD.cpp [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        string health_amount;
```
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: hope you like this one :dancing_women:

File: src/gamelogic/HUD.cpp [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        if (health < this->_amount_health_blocks) {
+            health_amount = "0";
+        }
+        health_amount += std::to_string(health);
+        _SDL_facade.draw_text(health_amount, FontType::alterebro_pixel, this->_health_text_color, this->_health_text_pos);
```
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: :dog:

File: src/gamelogic/HUD.cpp [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    void HUD::_draw_health_blocks(int health)
```
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: already done

File: src/gamelogic/HUD.cpp [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        int amount_of_blocks = this->_amount_health_blocks;
+        int set_off = 62;
```
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: no magic anymore :fireworks:

File: src/gamelogic/HUD.cpp [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+        for (int j = 0; j < this->_calculate_health_blocks(health); ++j) {
+            amount_of_blocks--;                   // for each lighter block we do not need an extra block
+            set_off = 62 + (j * (this->_health_block_width + 2)); // 2 is spacing
+            _SDL_facade.draw_rect({set_off, 433}, this->_health_block_width, this->_health_block_height, this->_health_block_full);
+        }
+
+        if (set_off != 62) {
```
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: done

File: src/gamelogic/HUD.cpp [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+            set_off += this->_health_block_width + 2;
+        }
+
+        for (int i = this->_amount_health_blocks - amount_of_blocks; i < this->_amount_health_blocks; ++i) {
+            set_off = 62 + (i * (this->_health_block_width + 2)); // 2 is spacing
```
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: no magic numbers anymore :fireworks:

File: src/gamelogic/HUD.cpp [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+            _SDL_facade.draw_rect({set_off, 433}, this->_health_block_width, this->_health_block_height, this->_health_block_empty);
+        }
+
+    }
+
+    void HUD::_draw_face()
```
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: done

File: src/gamelogic/HUD.cpp (hunk 5 above)
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: done

File: src/gamelogic/HUD.cpp (hunk 6 above)
Comment (SandervGeel) [inline src/gamelogic/HUD.cpp]: fixed it, don't know if it is better now?

File: src/gamelogic/HUD.hpp (hunk 1 above)
Comment (SandervGeel) [inline src/gamelogic/HUD.hpp]: deleted :kissing:

File: src/gamelogic/state/PauseState.cpp (hunk 3 above)
Comment (SandervGeel) [inline src/gamelogic/state/PauseState.cpp]: done

File: src/gamelogic/HUD.cpp [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+        for (int j = 0; j < this->_calculate_health_blocks(); ++j) {
+            amount_of_blocks--;                   // for each lighter block we do not need an extra block
+            xpos_draw_block = this->_set_off_health_bar + (j * (this->_health_block_width + this->_spacing_blocks));
+            this->_SDL_facade.draw_rect({xpos_draw_block, 433}, this->_health_block_width, this->_health_block_height, this->_health_block_full);
+        }
+
+        for (int i = this->_amount_health_blocks - amount_of_blocks; i < this->_amount_health_blocks; ++i) {
+            xpos_draw_block = this->_set_off_health_bar + (i * (this->_health_block_width + this->_spacing_blocks)); // 2 is spacing
+            this->_SDL_facade.draw_rect({xpos_draw_block, 433}, this->_health_block_width, this->_health_block_height, this->_health_block_empty);
```
Comment (jorgdebont) [inline src/gamelogic/HUD.cpp]: just one more magic number to remove mi feind

File: src/gamelogic/HUD.hpp [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+#include "../engine/interface/Drawable.hpp"
+#include "../engine/interface/Updatable.hpp"
+#include "../engine/SDLFacade.hpp"
+#include "../engine/PathUtil.hpp"
+#include "Player.hpp"
+#include <string>
+#include <iomanip>
+#include <sstream>
+
+#define HUD_PATH "res/HUDv2.bmp"
```
Comment (jorgdebont) [inline src/gamelogic/HUD.hpp]: define these as class variables

File: src/gamelogic/HUD.cpp [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+            this->_SDL_facade.draw_image(HUD_FACE_LEFT, this->_character_head_pos);
+        } else {
+            this->_SDL_facade.draw_image(HUD_FACE_RIGHT, this->_character_head_pos);
+        }
+    }
+
+    void HUD::draw()
+    {
+        //draw HUD
+        this->_SDL_facade.draw_image(HUD_PATH, this->_HUD_pos);
```
Comment (jorgdebont) [inline src/gamelogic/HUD.cpp]: this->

File: src/gamelogic/HUD.cpp [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+            xpos_draw_block = this->_set_off_health_bar + (i * (this->_health_block_width + this->_spacing_blocks)); // 2 is spacing
+            this->_SDL_facade.draw_rect({xpos_draw_block, this->_ypos_draw_block}, this->_health_block_width, this->_health_block_height, this->_health_block_empty);
+        }
+
+    }
+
+    void HUD::_draw_face()
+    {
+        if (this->_current_time % 4 > 1) {
+            this->_SDL_facade.draw_image(HUD_FACE_FRONT, this->_character_head_pos);
```
Comment (jorgdebont) [inline src/gamelogic/HUD.cpp]: this->

**Pull request merged**